crates/y-sweet-worker/src/s3_store.rs

```diff
@@ -0,0 +1,133 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use reqwest::{Client, StatusCode};
+use rusty_s3::{Bucket, Credentials, S3Action};
+use std::{cell::OnceCell, time::Duration};
+use time::OffsetDateTime;
+use y_sweet_core::store::Store;
+
+const PRESIGNED_URL_DURATION_SECONDS: u64 = 60 * 60;
+pub struct S3Store {
+    bucket: Bucket,
+    init_bucket: OnceCell<()>,
+    credentials: Credentials,
+    prefix: Option<String>,
+    presigned_url_duration: Duration,
+}
+
+impl S3Store {
+    pub fn new(
+        region: String,
+        bucket_name: String,
+        prefix: Option<String>,
+        aws_access_key_id: String,
+        aws_secret: String,
+    ) -> Self {
+        let credentials = Credentials::new(aws_access_key_id, aws_secret);
+        let endpoint = format!("https://s3.dualstack.{}.amazonaws.com", region)
+            .parse()
+            .expect("endpoint is a valid Url");
+        let path_style = rusty_s3::UrlStyle::VirtualHost;
+        let bucket = Bucket::new(endpoint, path_style, bucket_name, region)
+            .expect("Url has a valid scheme and host");
+
+        let presigned_url_duration = Duration::from_secs(PRESIGNED_URL_DURATION_SECONDS);
+        S3Store {
+            bucket,
+            init_bucket: OnceCell::new(),
+            credentials,
+            prefix,
+            presigned_url_duration,
+        }
+    }
+
+    async fn init_bucket(&self) -> Result<()> {
+        if self.init_bucket.get().is_none() {
+            let action = self.bucket.create_bucket(&self.credentials);
```

Comment:
We should never create a bucket; we should just raise an error if the bucket passed in does not exist.

Reply:
Oh, ok :+1: